src/mapreduce.jl

```diff
@@ -0,0 +1,276 @@
+    # does not support other indexing well
+    all_inds  = eachindex(IndexLinear(), srcs...)
+    batches   = Batches(all_inds, batch_size)
+    sample_inds = batches[1]
+    nt = Threads.nthreads()
+    arena_dst_view  = [_RollingCutOut(dst, sample_inds) for _ in 1:nt]
+    arena_src_views = create_arena_src_views(srcs, sample_inds)
+    return MapWorkspace(f, batches, arena_dst_view, arena_src_views)
+end
+
+@noinline function tmap!(f, dst, srcs::AbstractArray...;
+                        batch_size=default_batch_size(length(dst)))
+    isempty(first(srcs)) && return dst
+    w = prepare(tmap!, f, dst, srcs, batch_size=batch_size)
+    run!(w)
+    dst
+end
+
+function tmap(f, srcs::AbstractArray...; 
+             batch_size=default_batch_size(length(first(srcs)))
+            )
+    g = Base.Generator(f,srcs...)
+    T = Base.@default_eltype(g)
+    dst = similar(first(srcs), T)
+    tmap!(f, dst, srcs...; batch_size=batch_size)
+end
+
+############################## tmapreduce(like) ##############################
+struct Reduction{O}
+    op::O
+end
+(red::Reduction)(f, srcs...) = Base.mapreduce(f, red.op, srcs...)
+
+struct MapReduceWorkspace{R,F,B,V,OA<:OffsetArray}
+    reduction::R
+    f::F
+    batches::Batches{B}
+    arena_src_views::V
+    batch_reductions::OA
+end
+
+struct NoInit end
+
+function create_reduction(::typeof(tmapreduce), op)
+    Reduction(op)
+end
+
+function prepare(::typeof(tmapreduce), f, op, srcs; init, batch_size::Int)
+    red = create_reduction(tmapreduce, op)
+    w = prepare_mapreduce_like(red, f, srcs, init, batch_size=batch_size)
+    return w
+end
+
+function tmapreduce(f, op, srcs::AbstractArray...;
+                   init=NoInit(),
+                   batch_size= default_batch_size(length(first(srcs)))
+                  )
+    if isempty(first(srcs))
+        if init isa NoInit
+            return Base.mapreduce(f, op, srcs...)
+        else
+            return Base.mapreduce(f, op, srcs..., init=init)
+        end
+    end
+    w = prepare(tmapreduce, f, op, srcs, init=init, batch_size=batch_size)
+    run!(w)
+end
+
+function treduce(op, srcs::AbstractArray...; kw...)
+    tmapreduce(identity, op, srcs...; kw...)
+end
+
+
+for red in SYMBOLS_MAPREDUCE_LIKE
+    tred = tname(red)
+    @eval function $tred end
+
+    @eval function prepare(::typeof($tred), f, srcs; batch_size::Int)
+        base_red = Base.$red
+        prepare_mapreduce_like(base_red, f, srcs, batch_size=batch_size)
+    end
+
+    @eval function $tred(f, src::AbstractArray;
+                        batch_size=default_batch_size(length(src)))
+        isempty(src) && return Base.$red(f, src)
+        srcs = (src,)
+        w = prepare($tred, f, srcs, batch_size=batch_size)
+        run!(w)
+    end
+    @eval $tred(src; kw...) = $tred(identity, src; kw...)
+end
+
+function prepare_mapreduce_like(red, f, srcs, init=NoInit(); batch_size::Int)
+    all_inds  = eachindex(IndexLinear(), srcs...)
```

Comment:
I don't quite understand why we explicitly use `IndexLinear()` here, why not let `eachindex` decide. Is there any case where `eachindex(IndexLinear(), srcs...)` differs from `eachindex(srcs...)`?

Reply:
Good point, this is probably not needed. I will remove it.

Reply:
Ah I remember `IndexLinear` was choosen for the same reason as [above](https://github.com/mohamed82008/KissThreading.jl/pull/23#discussion_r308531195).

Reply:
I don't understand what you mean by the above comment. May you give me a case where `eachindex(IndexLinear(), srcs...)` and `eachindex(srcs...)` give different outputs?

Reply:
```julia
julia> eachindex(view(randn(2,2), 2:2, 1:1))
1×1 CartesianIndices{2,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}}}:
 CartesianIndex(1, 1)
```